tuner/tuner/libtuner_test.py

```diff
@@ -188,60 +188,220 @@ def test_get_compilation_success_rate():
+    ]
+    assert not libtuner.are_baseline_devices_unique(baseline_results)
+
+
+def test_detect_baseline_regression():
+    first_baseline = [
+        libtuner.BenchmarkResult(0, 1000.0, "hip://0"),
+        libtuner.BenchmarkResult(0, 2000.0, "hip://1"),
+    ]
+    second_baseline = [
+        libtuner.BenchmarkResult(0, 1100.0, "hip://0"),
+        libtuner.BenchmarkResult(0, 1900.0, "hip://1"),
+    ]
+    regression_devices = libtuner.detect_baseline_regression(
+        first_baseline, second_baseline
     )
-    assert best_results[0].candidate_id == 4
-    assert best_results[1].candidate_id == 3
-    assert best_results[2].candidate_id == 2
+    assert regression_devices == ["hip://0"]
 
+    first_baseline = [
+        libtuner.BenchmarkResult(0, 1000.0, "hip://0"),
+        libtuner.BenchmarkResult(0, 2000.0, "hip://1"),
+    ]
+    second_baseline = [
+        libtuner.BenchmarkResult(0, 1000.0, "hip://0"),
+        libtuner.BenchmarkResult(0, 1000.0, "hip://1"),
+    ]
+    regression_devices = libtuner.detect_baseline_regression(
+        first_baseline, second_baseline
+    )
+    assert regression_devices == []
 
-def test_enum_collision():
-    from iree.compiler.dialects import linalg, vector, iree_gpu, iree_codegen, iree_input  # type: ignore
+    first_baseline = [
+        libtuner.BenchmarkResult(0, 1000.0, "hip://0"),
+        libtuner.BenchmarkResult(0, 2000.0, "hip://1"),
+    ]
+    second_baseline = [
+        libtuner.BenchmarkResult(0, 1100.0, "hip://0"),
+    ]
+    regression_devices = libtuner.detect_baseline_regression(
+        first_baseline, second_baseline
+    )
+    assert regression_devices == ["hip://0"]
+
+
+def test_add_run():
+    handler = libtuner.BaselineResultHandler()
+    baseline = [
+        libtuner.BenchmarkResult(1, 0.5, "hip://0"),
+        libtuner.BenchmarkResult(2, math.inf, "hip://1"),
+        libtuner.BenchmarkResult(3, 0.7, "hip://0"),
+    ]
+    handler.add_run(baseline)
+
+    assert handler.runs["hip://0"] == [0.5, 0.7]
+    assert handler.runs["hip://1"] == []
+
+
+def test_num_successful_runs():
+    handler = libtuner.BaselineResultHandler()
+    assert handler.num_successful_runs("hip://0") == 0
+    assert handler.num_successful_runs("hip://1") == 0
+
+    baseline = [
+        libtuner.BenchmarkResult(1, 0.5, "hip://0"),
+        libtuner.BenchmarkResult(2, math.inf, "hip://1"),
+        libtuner.BenchmarkResult(3, 0.7, "hip://0"),
+        libtuner.BenchmarkResult(4, math.nan, "hip://1"),
+    ]
+    handler.add_run(baseline)
+    assert handler.num_successful_runs("hip://0") == 2
+    assert handler.num_successful_runs("hip://1") == 0
+
+    additional_baseline = [
+        libtuner.BenchmarkResult(5, 1.2, "hip://2"),
+        libtuner.BenchmarkResult(6, 0.9, "hip://2"),
+    ]
+    handler.add_run(additional_baseline)
+    assert handler.num_successful_runs("hip://2") == 2
+    assert handler.num_successful_runs("hip://3") == 0
+
+
+def test_get_average_result():
```

Comment:
We can test multiple methods in one function to avoid repeating the same setup code. Not every function has to have its own test function.